ecosystem/typescript/sdk/src/plugins/fungible_asset_client.ts

```diff
@@ -0,0 +1,113 @@
+import { AptosAccount } from "../account";
+import { RawTransaction } from "../aptos_types";
+import * as Gen from "../generated/index";
+import { OptionalTransactionArgs, Provider } from "../providers";
+import { TransactionBuilderRemoteABI } from "../transaction_builder";
+import { MaybeHexString, HexString } from "../utils";
+
+export class FungibleAssetClient {
+  provider: Provider;
+
+  readonly assetType: string = "0x1::fungible_asset::Metadata";
+
+  /**
+   * Creates new FungibleAssetClient instance
+   *
+   * @param provider Provider instance
+   */
+  constructor(provider: Provider) {
+    this.provider = provider;
+  }
+
+  /**
+   * Transfer an asset amount from the sender primary_fungible_store to the recipient primary_fungible_store
+   *
+   * Use this method to transfer any fungible asset including fungible token.
+   *
+   * @param sender The sender account
+   * @param assetAddress The fungible asset address.
+   * For example if you’re transferring USDT this would be the USDT address
+   * @param recipient Recipient address
+   * @param amount Number of assets to transfer
+   * @param assetType (optional) The fungible asset type - default to `0x1::fungible_asset::Metadata`
+   * @returns The hash of the transaction submitted to the API
+   */
+  async transferAmount(
+    sender: AptosAccount,
+    assetAddress: MaybeHexString,
+    recipient: MaybeHexString,
+    amount: number | bigint,
+    assetType?: string,
+    extraArgs?: OptionalTransactionArgs,
+  ): Promise<string> {
+    const rawTxn = await this.generateTransferAmount(sender, assetAddress, recipient, amount, assetType, extraArgs);
+    const txnHash = await this.submit(sender, rawTxn);
+    return txnHash;
+  }
+
+  /**
+   * Get the balance of an account's fungible asset.
+   *
+   * @param account Account that you want to get the balance of.
+   * @param assetAddress The fungible asset address you want to check the balance of
+   * @param assetType (optional) The fungible asset type - default to `0x1::fungible_asset::Metadata`
+   * @returns Promise that resolves to the balance
+   */
+  async balance(account: MaybeHexString, assetAddress: MaybeHexString, assetType?: string): Promise<bigint> {
+    const payload: Gen.ViewRequest = {
+      function: "0x1::primary_fungible_store::balance",
```

Comment:
Is there something particularly complicated about the underlying move structs explaining why we need to use a view function for this? I guess it needs to fetch the metadata by following an "object pointer"?

If not and it's just a single resource read, given reading a coin / fungible asset is a pretty common operation, it could be better to do it via the API normally.

Reply:
It should be supported natively in API at object level since we don't have indexer inside node and the ownership is not that intuitive as coin v1. So you know the balance of someone you'll query indexer first to get the fungible store object address and then call the api.